primitives/core/src/bandersnatch.rs

```diff
@@ -0,0 +1,1010 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! VRFs backed by [Bandersnatch](https://neuromancer.sk/std/bls/Bandersnatch),
+//! an elliptic curve built over BLS12-381 scalar field.
+//!
+//! The Bandersnatch curve can be represented in twisted Edwards coordinates, allowing
+//! efficieny inside zk-SNARKS circuits.
+
+#[cfg(feature = "std")]
+use crate::crypto::Ss58Codec;
+use crate::crypto::{
+	ByteArray, CryptoType, CryptoTypeId, Derive, Public as TraitPublic, UncheckedFrom, VrfPublic,
+};
+#[cfg(feature = "full_crypto")]
+use crate::crypto::{DeriveError, DeriveJunction, Pair as TraitPair, SecretStringError, VrfSecret};
+
+use bandersnatch_vrfs::CanonicalSerialize;
+#[cfg(feature = "full_crypto")]
+use bandersnatch_vrfs::SecretKey;
+use codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+
+use sp_runtime_interface::pass_by::PassByInner;
+use sp_std::{boxed::Box, vec::Vec};
+
+/// Identifier used to match public keys against bandersnatch-vrf keys.
+pub const CRYPTO_ID: CryptoTypeId = CryptoTypeId(*b"bs38");
+
+/// Context used to produce
+#[cfg(feature = "full_crypto")]
+pub const SIGNING_CTX: &[u8] = b"SigningContext";
+
+#[cfg(feature = "full_crypto")]
+const SEED_SERIALIZED_LEN: usize = 32;
+
+// Edwards form sizes
+// @burdges @swasilyev: currently ring-proof is using SHORT-WEIERSTRASS
+// I had to temporary patch bandersnatch_vrfs crate to use SW instrad of ED...
+// (@davxy: probably we'll use ED form)
+// const PUBLIC_SERIALIZED_LEN: usize = 32;
+// const SIGNATURE_SERIALIZED_LEN: usize = 64;
+
+// Short-Weierstrass form sizes (TEMPORARY)
+const PUBLIC_SERIALIZED_LEN: usize = 33;
+const SIGNATURE_SERIALIZED_LEN: usize = 65;
```

Comment:
@burdges @swasilyev  `bandersnatch_vrfs` was originally using `ED` form while `ring-proof` requires `SW` form.

I temporary changed `bandersnatch_vrfs` to use `SW` as well. But I suspect we need to change `ring-proof` to use `ED` instead?

Reply:
We'll keep this crate primarily SW I think.  I asked @achimcc to implement elligator and the map from edwards to SW, but this is to optimize the hash-to-curve, and to be compliant with what the IRTF draft standard implies.